Sources/TensorFlow/Operators/NN.swift

```diff
@@ -138,434 +137,425 @@ public extension Padding {
+    return (value, { v in
+        (v.conv2DBackpropInput(
+            shape: x.shapeTensor, filter: filter, strides: strides, padding: padding),
+         v.conv2DBackpropFilter(
+            input: x, filterSizes: filter.shapeTensor, strides: strides, padding: padding))
+    })
+}
 
-    @inlinable
-    internal func _vjpConvolved3D(
-        filter: Tensor,
-        strides: (Int, Int, Int, Int, Int),
-        padding: Padding
-    ) -> (Tensor, (Tensor) -> (Tensor, Tensor)) {
-        let value = convolved3D(withFilter: filter, strides: strides,
-                                padding: padding)
-        return (value, { v in
-            return (
-                v.conv3DBackpropInput(
-                    shape: self.shapeTensor, filter: filter,
-                    strides: strides, padding: padding
-                ),
-                v.conv3DBackpropFilter(
-                    input: self, filterSizes: filter.shapeTensor,
-                    strides: strides, padding: padding
-                )
-            )
-        })
-    }
+/// TensorFlow builtin conv3d gradient helper for the input.
+@differentiable(wrt: (x, filter), vjp: _vjpConv3DBackpropInput)
+@usableFromInline
+func conv3DBackpropInput<Scalar: TensorFlowFloatingPoint>(
+    _ x: Tensor<Scalar>,
+    shape: Tensor<Scalar><Int32>,
+    filter: Tensor<Scalar>,
+    strides: (Int, Int, Int, Int, Int),
+    padding: Padding
+) -> Tensor<Scalar> {
+    return Raw.conv3DBackpropInputV2(
+        inputSizes: shape,
+        filter: filter,
+        outBackprop: x,
+        strides: [Int32(strides.0), Int32(strides.1), Int32(strides.2),
+                  Int32(strides.3), Int32(strides.4)],
+        padding: padding.raw)
+}
 
-    @inlinable
-    internal func _vjpMaxPooled2D(
-        kernelSize: (Int, Int, Int, Int),
-        strides: (Int, Int, Int, Int),
-        padding: Padding
-    ) -> (Tensor, (Tensor) -> Tensor) {
-        // TODO: Currently this is not higher order differentiable. Redefine in
-        // closed form.
-        let value = maxPooled2D(kernelSize: kernelSize, strides: strides, padding: padding)
-        return (value, { v in
-            Raw.maxPoolGradV2(
-                origInput: self,
-                origOutput: value,
-                grad: v,
-                ksize: Tensor<Int32>([Int32(kernelSize.0), Int32(kernelSize.1),
-                                      Int32(kernelSize.2), Int32(kernelSize.3)]),
-                strides: Tensor<Int32>([Int32(strides.0), Int32(strides.1),
-                                        Int32(strides.2), Int32(strides.3)]),
-                padding: padding.raw)
-        })
-    }
+/// TensorFlow builtin conv3d gradient helper for the filter.
+@differentiable(wrt: (x, input), vjp: _vjpConv3DBackpropFilter)
+@usableFromInline
+func conv3DBackpropFilter<Scalar: TensorFlowFloatingPoint>(
+    _ x: Tensor<Scalar>,
+    input: Tensor<Scalar>,
+    filterSizes: Tensor<Scalar><Int32>,
```

Comment:
```suggestion
    filterSizes: Tensor<Int32>,
```
This change is incorrect. Please carefully examine each occurrence when doing a find-and-replace, and please make sure the code compiles locally on your machine.